packages/libs/pactjs-generator/src/lexer.ts

```diff
@@ -0,0 +1,243 @@
+import type { Lexer } from 'moo';
+import * as moo from 'moo';
+
+/**
+ * @internal
+ */
+export const lexer: Lexer = moo.compile({
+  TYPED_IDEN: /[a-zA-Z-_0-9]+:[a-zA-Z-_{}.0-9]+/,
+  name: {
+    match: /[a-zA-Z-_*]+/,
+    type: moo.keywords({
+      KW_BLESS: 'bless',
+      KW_DEFCONST: 'defconst',
+      KW_DEFPACT: 'defpact',
+      KW_DEFSCHEMA: 'defschema',
+      KW_DEFTABLE: 'deftable',
+      KW_DEFUN: 'defun',
+      KW_DEFCAP: 'defcap',
+      KW_FALSE: 'false',
+      KW_IMPLEMENTS: 'implements',
+      KW_INTERFACE: 'interface',
+      KW_LET: 'let',
+      KW_LET_STAR: 'let*',
+      KW_LAMBDA: 'lambda',
+      KW_MODULE: 'module',
+      KW_STEP: 'step',
+      KW_STEP_WITH_ROLLBACK: 'step-with-rollback',
+      KW_TRUE: 'true',
+      KW_USE: 'use',
+      KW_COND: 'cond',
+      KW_WITH_CAPABILITY: 'with-capability',
+    }),
+  },
+  postEscapedNewline: /\s*[\\|\\\\]/,
+  WS: /[ \t]+/,
+  lparen: '(',
+  rparen: ')',
+  lsquare: '[',
+  rsquare: ']',
+  lcurly: '{',
+  rcurly: '}',
+  assign: ':=',
+
+  comma: ',',
+  string: { match: /"(?:[^"\\]|\\.|\\\n)*"/, lineBreaks: true },
+  newlineEscape: {
+    match: /[\\|\\\\][\n]\s*/,
+    lineBreaks: true,
+  },
+  singleQuote: /'/,
+  doubleQuote: /"/,
+  plus: '+',
+  number: /-?\d+/,
+  colon: ':',
+  equals: '=',
+  comment: /;;?.*/,
+  module: /\'[a-zA-Z_-]+/,
+  guard: /@[a-zA-Z-_]+/,
+  gt: '>',
+  lt: '<',
+  float: /-?\d+\.\d+/,
+  dot: '.',
+  NOT_IS: /!=/,
+  NL: { match: /\n/, lineBreaks: true },
+});
+
+/**
+ * @internal
+ */
+export interface ILocation {
+  line: number;
+  col: number;
+}
+/**
+ * @internal
+ */
+export type Arg = { name: string; type: string | undefined } & ILocation;
+/**
+ * @internal
+ */
+export type Defun = {
+  method: string;
+  returnType: string;
+  args: Record<string, Arg>;
+} & ILocation;
+/**
+ * @internal
+ */
+export interface IState {
+  stack: string[];
+  moduleName: string;
+  defunName: string;
+}
+
+/**
+ * @internal
+ */
+export type Module = {
+  name: string;
+  defuns: Record<string, Defun>;
+} & ILocation;
+/**
+ * @internal
+ */
+export type Output = Record<string, Module>;
+
+/**
+ * @internal
+ */
+export interface ILogger {
+  (msg: object | string, ...args: Array<object | string>): void;
+}
+
+/**
+ * @internal
+ */
+export function getModuleAndMethods(contract: string, logger: ILogger): Output {
+  lexer.reset(contract);
+
+  let token: moo.Token | undefined = undefined;
+  const state: IState = { stack: [], moduleName: '', defunName: '' } as IState;
+
+  const output: Output = {} as Output;
+  while ((token = lexer.next())) {
+    logger(token);
+    switch (token.type) {
+      case 'KW_MODULE':
+        state.stack.push('module');
+        logger('push module');
```

Comment:
Shouldn't we also log `token.type` here to get more specific logs?

Reply:
It's logging the full object at the start of the `while` loop